plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -1041,6 +1043,46 @@ else if (column.getId() == TRINO_MERGE_PARTITION_DATA) {
         }
     }
 
+    private static Map<Integer, org.apache.parquet.schema.Type> createParquetIdToFieldMapping(MessageType fileSchema)
+    {
+        ImmutableMap.Builder<Integer, org.apache.parquet.schema.Type> builder = ImmutableMap.builder();
+        addParquetIdToFieldMapping(fileSchema, builder);
+        return builder.buildOrThrow();
+    }
+
+    private static void addParquetIdToFieldMapping(org.apache.parquet.schema.Type type, ImmutableMap.Builder<Integer, org.apache.parquet.schema.Type> builder)
+    {
+        if (type.getId() != null) {
+            builder.put(type.getId().intValue(), type);
+        }
+        if (type instanceof PrimitiveType) {
+            // Nothing else to do
+        }
+        else if (type instanceof GroupType groupType) {
+            for (org.apache.parquet.schema.Type field : groupType.getFields()) {
+                addParquetIdToFieldMapping(field, builder);
+            }
+        }
+        else {
+            throw new IllegalStateException("Unsupported field type: " + type);
+        }
+    }
+
+    private static MessageType getMessageType(List<IcebergColumnHandle> regularColumns, String fileSchemaName, Map<Integer, org.apache.parquet.schema.Type> parquetIdToField)
```

Comment:
This is an adaptation of the code taken from: https://github.com/trinodb/trino/blob/c0134f1db03eed3f080bdf4b8894fdd7df0fdcd7/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java#L317-L331

Reply:
I was trying to convert `IcebergColumnHandle` to `HiveColumnHandle` and make use of `ParquetPageSourceFactory#getParquetMessageType`.

it seems that Hive does not support `timestamp(3) with time zone` type.

```
io.trino.spi.TrinoException: Unsupported Hive type: timestamp(3) with time zone
	at io.trino.plugin.hive.util.HiveTypeTranslator.toTypeInfo(HiveTypeTranslator.java:174)
	at io.trino.plugin.hive.HiveType.toHiveType(HiveType.java:233)
	at io.trino.plugin.iceberg.IcebergColumnHandle.toHiveColumnHandle(IcebergColumnHandle.java:308)
```
I will go ahead and review the current implementation. Thanks!